src/compiler.jl

```diff
@@ -1012,19 +1012,109 @@ function wait_rev(B::LLVM.API.LLVMBuilderRef, OrigCI::LLVM.API.LLVMValueRef, gut
     return nothing
 end
 
-function arraycopy_fwd(B::LLVM.API.LLVMBuilderRef, OrigCI::LLVM.API.LLVMValueRef, gutils::API.EnzymeGradientUtilsRef, normalR::Ptr{LLVM.API.LLVMValueRef}, shadowR::Ptr{LLVM.API.LLVMValueRef}, tapeR::Ptr{LLVM.API.LLVMValueRef})::Cvoid
-    emit_error(LLVM.Builder(B), "Enzyme: Not yet implemented forward for jl_array_copy")
+function arraycopy_common(fwd, B, orig, origArg, gutils)
+    # size_t len = jl_array_len(ary);
+    # size_t elsz = ary->elsize;
+    # memcpy(new_ary->data, ary->data, len * elsz);
+	# JL_EXTENSION typedef struct {
+	# 	JL_DATA_TYPE
+	# 	void *data;
+	# #ifdef STORE_ARRAY_LEN
+	# 	size_t length;
+	# #endif
+	# 	jl_array_flags_t flags;
+	# 	uint16_t elsize;  // element size including alignment (dim 1 memory stride)
+
+	tt = TypeTree(API.EnzymeGradientUtilsAllocAndGetTypeTree(gutils, orig))
```

Comment:
Can we define actual Julia functions for these, consistent with the rest of typetree.jl?